compiler/damlc/lib/DA/Cli/Damlc/Command/MultiIde.hs

```diff
@@ -0,0 +1,519 @@
+parseCustomResult name =
+  fmap $ either (\err -> error $ "Failed to parse response of " <> name <> ": " <> err) id 
+    . Aeson.parseEither Aeson.parseJSON 
+
+-- Handlers
+
+subIDEMessageHandler :: MultiIdeState -> IO () -> SubIDE -> B.ByteString -> IO ()
+subIDEMessageHandler miState unblock ide bs = do
+  debugPrint "Called subIDEMessageHandler"
+  -- BSC.hPutStrLn stderr bs
+
+  -- Decode a value, parse
+  let val :: Aeson.Value
+      val = er "eitherDecode" $ Aeson.eitherDecodeStrict bs
+  mMsg <- either error id <$> parseServerMessageWithTracker (fromClientMethodTrackerVar miState) (ideHomeDirectory ide) val
+
+  -- Adds the various prefixes needed for from server messages to not clash with those from other IDEs
+  mPrefixedMsg <-
+    mapM 
+      ( addProgressTokenPrefixToServerMessage (serverCreatedProgressTokensVar miState) (ideHomeDirectory ide) (ideMessageIdPrefix ide)
+      . addLspPrefixToServerMessage ide
+      )
+      mMsg
+
+  forM_ mPrefixedMsg $ \msg -> do
+    -- If its a request (builtin or custom), save it for response handling.
+    putServerReq (fromServerMethodTrackerVar miState) (ideHomeDirectory ide) msg
+
+    debugPrint "About to thunk message"
+    case msg of
+      LSP.FromServerRsp LSP.SInitialize LSP.ResponseMessage {_result} -> do
+        debugPrint "Got initialization reply, sending initialized and unblocking"
+        -- Dangerous call here is acceptable as this only happens while the ide is booting, before unblocking
+        sendSubIDE ide $ LSP.FromClientMess LSP.SInitialized $ LSP.NotificationMessage "2.0" LSP.SInitialized (Just LSP.InitializedParams)
+        unblock
+      LSP.FromServerRsp LSP.SShutdown _ -> handleExit miState ide
+
+      LSP.FromServerRsp (LSP.SCustomMethod "daml/tryGetDefinition") LSP.ResponseMessage {_id, _result} -> do
+        debugPrint "Got tryGetDefinition response, handling..."
+        let parsedResult = parseCustomResult @(Maybe TryGetDefinitionResult) "daml/tryGetDefinition" _result
+            reply :: Either LSP.ResponseError (LSP.ResponseResult 'LSP.TextDocumentDefinition) -> IO ()
+            reply rsp = do
+              debugPrint $ "Replying directly to client with " <> show rsp
+              sendClient miState $ LSP.FromServerRsp LSP.STextDocumentDefinition $ LSP.ResponseMessage "2.0" (castLspId <$> _id) rsp
+            replyLocations :: [LSP.Location] -> IO ()
+            replyLocations = reply . Right . LSP.InR . LSP.InL . LSP.List
+        case parsedResult of
+          Left err -> reply $ Left err
+          Right Nothing -> replyLocations []
+          Right (Just (TryGetDefinitionResult loc Nothing)) -> replyLocations [loc]
+          Right (Just (TryGetDefinitionResult loc (Just name))) -> do
+            debugPrint $ "Got name in result! Backup location is " <> show loc
+            let mHome = Map.lookup (tgdnPackageUnitId name) $ multiPackageMapping miState
+            case mHome of
+              Nothing -> replyLocations [loc]
```

Comment:
Might be worth commenting the difference between good path on 354 and bad path here